### PR TITLE
Expand spring-security fuzzing

### PR DESCRIPTION
**Pull request by CheeseHunter117**
Added fuzz targets for:
- crypto.util.EncodingUtils.concatenate
- crypto.codec.Hex.encode and crypto.codec.Hex.decode
- crypto.codec.Utf8.encode and crypto.codec.Utf8.decode
- provisioning.InMemoryUserDetailsManager.changePassword

### CODE REVIEW

**Comment by CheeseHunter117** [comment]
@onionpsy Please review

File: projects/spring-security/EncodingUtilsConcatenateFuzzer.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+public class EncodingUtilsConcatenateFuzzer {
+    public static void fuzzerTestOneInput(FuzzedDataProvider data) {
+        final byte[][] arrayOfByteArrays = getArrayOfByteArrays(data);
+
+        EncodingUtils.concatenate(arrayOfByteArrays);
+    }
+
+    // Constants to reduce cases of fuzzer running out of memory
+    // Values chosen without heuristics or logic
```
Comment (onionpsy) [inline projects/spring-security/EncodingUtilsConcatenateFuzzer.java:13]: I would remove this comment

File: projects/spring-security/HexFuzzer.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+public class HexFuzzer {
+    public static void fuzzerTestOneInput(byte[] data) {
+        final byte[] initialByteArray = data;
+        final char[] encodedChars;
+
+        try {
+            encodedChars = Hex.encode(initialByteArray);
+
+            if (! initialByteArray.toString().equals(Hex.decode(encodedChars.toString()))) {
+                throw new FuzzerSecurityIssueHigh("Hex value has changed during encoding and decoding");
```
Comment (onionpsy) [inline projects/spring-security/HexFuzzer.java:16]: `FuzzerSecurityIssueHigh` is probably a bit too much, I would set to low or medium

File: projects/spring-security/InMemoryUserDetailsManagerChangePasswordFuzzer.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.security.provisioning.InMemoryUserDetailsManager;
+
+public class InMemoryUserDetailsManagerChangePasswordFuzzer {
+    private final static String USERNAME = "admin";
+    private final static String PASSWORD = "secret";
+    private final static String USER_ROLE = "ADMIN";
+    private static final List<GrantedAuthority> AUTHORITIES = AuthorityUtils.createAuthorityList(USER_ROLE);
+
+    // Values chosen without heuristics or logic
```
Comment (onionpsy) [inline projects/spring-security/InMemoryUserDetailsManagerChangePasswordFuzzer.java:21]: Same here

File: projects/spring-security/InMemoryUserDetailsManagerChangePasswordFuzzer.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        // create all the objects needed for fuzzing the InMemoryUserDetailsManager
+        final User user = new User(USERNAME, PASSWORD, AUTHORITIES);
+        final InMemoryUserDetailsManager userDetailsManager = new InMemoryUserDetailsManager(user);
+
+        // set the SecurityContext
+        // both options below make it so that InMemoryUserDetailsManager.changePassword(old, new) does not actually check the old password
+        SecurityContextHolder.getContext().setAuthentication(
+            UsernamePasswordAuthenticationToken.authenticated(USERNAME, PASSWORD, AUTHORITIES));
+        // SecurityContextHolder.getContext().setAuthentication(UsernamePasswordAuthenticationToken.unauthenticated(USERNAME, PASSWORD));
```
Comment (onionpsy) [inline projects/spring-security/InMemoryUserDetailsManagerChangePasswordFuzzer.java:42]: Please removed commented code

File: projects/spring-security/InMemoryUserDetailsManagerChangePasswordFuzzer.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        try {
+            userDetailsManager.changePassword(generatedPassword01, generatedPassword02);
+
+            // check if the password was successfully changed
+            final String finalPassword = userDetailsManager.loadUserByUsername(USERNAME).getPassword();
+            if (PASSWORD.equals(finalPassword)) {
+                throw new FuzzerSecurityIssueHigh("Password was not changed to '" + finalPassword + "'");
+            }
+        } catch (UsernameNotFoundException err) {
+            throw new FuzzerSecurityIssueHigh("The user disappeared from the InMemoryUserDetailsManager");
```
Comment (onionpsy) [inline projects/spring-security/InMemoryUserDetailsManagerChangePasswordFuzzer.java:53]: Probably more a bug than a security if it reach that point. We don't have a "bug" exception in jazzer so I would just throw a security low here

File: projects/spring-security/Utf8Fuzzer.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+public class Utf8Fuzzer {
+    public static void fuzzerTestOneInput(FuzzedDataProvider data) {
+        final String initialString = data.consumeString(Integer.MAX_VALUE);
+        final byte[] encodedBytes;
+
+        try {
+            encodedBytes = Utf8.encode(initialString);
+
+            if (! initialString.equals(Utf8.decode(encodedBytes))) {
+                throw new FuzzerSecurityIssueHigh("Utf8 value has changed during encoding and decoding");
```
Comment (onionpsy) [inline projects/spring-security/Utf8Fuzzer.java:17]: Same here, high is a bit too much

**Review by onionpsy — CHANGES_REQUESTED** [review]

**Pull request merged**